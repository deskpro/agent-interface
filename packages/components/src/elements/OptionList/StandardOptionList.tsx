import * as React from "react";
import OptionList from "./OptionList";
import Menu from "../Menu/Menu";

export type OptionItemSettings = {
  name: React.Key;
  text: string;
};

export type OptionItemType = {
  name: React.Key;
  title: string;
  disabled?: boolean;
  checked?: boolean;
  settings?: OptionItemSettings[];
};

export type StandardOptionListProps = {
  name: string;
  className?: string;
  items: OptionItemType[];
  sortable?: boolean;
  onOrderChange?: (
    itemName: React.Key,
    oldPosition: number,
    newPosition: number
  ) => void;
  checkable?: boolean;
  onCheck?: (itemName: React.Key, checked: boolean) => void;
  onSettingChange?: (settingName: React.Key, checked: boolean) => void;
};

const StandardOptionList: React.FC<StandardOptionListProps> = ({
  items,
  onCheck,
  onSettingChange,
  ...props
}) => (
  <OptionList {...props}>
    {items.map(({ settings, ...item }) => (
      <OptionList.Item
        {...item}
        key={item.name}
        onCheck={onCheck}
        renderGearMenu={
          settings
            ? menuProps => (
                <Menu
                  {...menuProps}
                  isCheckable
                  onItemCheck={onSettingChange}
                  checkedItems={[settings[0].name]}
                >
                  {settings.map(setting => (
                    <Menu.MenuItem {...setting} key={setting.name} />
                  ))}
                </Menu>
              )
            : undefined
        }
      />
    ))}
  </OptionList>
);

export default StandardOptionList;
