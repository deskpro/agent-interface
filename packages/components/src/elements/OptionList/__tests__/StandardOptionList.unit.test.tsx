import * as React from "react";
import { render, fireEvent } from "react-testing-library";

import StandardOptionList from "../StandardOptionList";

const items = [
  { name: "user", title: "User", disabled: true },
  { name: "agent_team", title: "Agent Team" },
  { name: "labels", title: "Labels" },
  { name: "id", title: "ID" },
  {
    name: "created",
    title: "Created",
    settings: [
      { name: "created.date_only", text: "Date" },
      { name: "created.date_time", text: "Date/Time" },
      { name: "created.time_ago", text: "Time Ago" }
    ]
  },
  {
    name: "waiting_time",
    title: "All waiting time",
    settings: [
      { name: "waiting_time.date_only", text: "Date" },
      { name: "waiting_time.date_time", text: "Date/Time" },
      { name: "waiting_time.time_ago", text: "Time Ago" }
    ]
  }
];

describe("<StandardOptionList />", () => {
  it("should match its snapshot", () => {
    const { container } = render(
      <StandardOptionList
        name="fields"
        items={items}
        checkable
        onCheck={jest.fn()}
        onOrderChange={jest.fn()}
        onSettingChange={jest.fn()}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
