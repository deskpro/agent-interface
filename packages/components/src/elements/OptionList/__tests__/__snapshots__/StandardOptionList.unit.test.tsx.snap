// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StandardOptionList /> should match its snapshot 1`] = `
<div
  class="dp-OptionList dp-Menu"
>
  <div
    class="dp-OptionList-item dp-Menu-linkItem dp-Uneditable"
  >
    <span
      class="dp-CustomCheckbox"
    >
      <input
        class="dp-ControlInput"
        disabled=""
        id="user"
        name="user"
        type="checkbox"
      />
      <label
        class=""
        for="user"
      />
    </span>
    User
  </div>
  <div
    class="dp-OptionList-item dp-Menu-linkItem"
  >
    <span
      class="dp-CustomCheckbox"
    >
      <input
        class="dp-ControlInput"
        id="agent_team"
        name="agent_team"
        type="checkbox"
      />
      <label
        class=""
        for="agent_team"
      />
    </span>
    Agent Team
  </div>
  <div
    class="dp-OptionList-item dp-Menu-linkItem"
  >
    <span
      class="dp-CustomCheckbox"
    >
      <input
        class="dp-ControlInput"
        id="labels"
        name="labels"
        type="checkbox"
      />
      <label
        class=""
        for="labels"
      />
    </span>
    Labels
  </div>
  <div
    class="dp-OptionList-item dp-Menu-linkItem"
  >
    <span
      class="dp-CustomCheckbox"
    >
      <input
        class="dp-ControlInput"
        id="id"
        name="id"
        type="checkbox"
      />
      <label
        class=""
        for="id"
      />
    </span>
    ID
  </div>
  <div
    class="dp-OptionList-item dp-Menu-linkItem"
  >
    <span
      class="dp-CustomCheckbox"
    >
      <input
        class="dp-ControlInput"
        id="created"
        name="created"
        type="checkbox"
      />
      <label
        class=""
        for="created"
      />
    </span>
    Created
    <span
      class="dp-OptionList-itemGearMenu"
    >
      <span
        class="dp-Icon Icon--settings"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="12"
          width="12"
        />
      </span>
    </span>
  </div>
  <div
    class="dp-OptionList-item dp-Menu-linkItem"
  >
    <span
      class="dp-CustomCheckbox"
    >
      <input
        class="dp-ControlInput"
        id="waiting_time"
        name="waiting_time"
        type="checkbox"
      />
      <label
        class=""
        for="waiting_time"
      />
    </span>
    All waiting time
    <span
      class="dp-OptionList-itemGearMenu"
    >
      <span
        class="dp-Icon Icon--settings"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="12"
          width="12"
        />
      </span>
    </span>
  </div>
</div>
`;
